Implement automatic client registration

rmiccoli opened the pull request:
This PR implements one of the client/Relying Party registration mechanisms defined by OpenID Federation.

Automatic Registration allows an RP to send Authentication Requests without a prior registration step with the OP. When the RP reaches the OP’s Authorization Endpoint, the OP automatically registers the RP if it can build and validate a trust chain for the RP - either by resolving the RP’s Entity Configuration from its `entity_id` or by using a provided `trust_chain` - and a common Trust Anchor exists.

At the Authorization Endpoint, in fact, the RP can use a Request Object by sending the `request` parameter, a signed JWT containing some required claims (`aud`, `client_id`, `iss`, `jti`) and optionally `iat` and `trust_chain`.

Note that the OP assigns the `client_id` to the RP, which must correspond to its `entity_id` (URL).


Reference: https://openid.net/specs/openid-federation-1_0.html#name-automatic-registration

rmiccoli committed:
Implement automatic client registration

by moving AuthZRequestFilter from MITREid to IAM

rmiccoli committed:
Add verification of the request object signature

rmiccoli committed:
Fix after rebase from develop

rmiccoli committed:
Add missing license

rmiccoli committed:
Fix some code smells

rmiccoli committed:
Fix other code smells

rmiccoli committed:
Validate URL before constructing it

rmiccoli committed:
Refactor methods to reduce complexity

rmiccoli committed:
Combine common logic into an abstract class

rmiccoli committed:
Fix code smell and move handler

rmiccoli committed:
Add info in OP Entity Configuration

in particular, support for automatic client
registration and no support for authentication
requests with request_uri

rmiccoli committed:
Add test

rmiccoli committed:
Add other tests

rmiccoli committed:
Refactor error responses

and add url validation

rmiccoli committed:
Replace mitreID client service

with IamClientRepository and other fixes

rmiccoli committed:
Fix client initialization

and remove unused method parameter

rmiccoli committed:
Simplify fetchEntityConfiguration method

and add tests

rmiccoli committed:
Delete the extra try

rmiccoli committed:
Refactor code

rmiccoli committed:
Simplify doFilter "brain" method

rmiccoli committed:
Add tests

rmiccoli committed:
Refactor code and add tests

rmiccoli committed:
Increase coverage

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

federicaagostini commented:
Remember also to run the [disableExpiredClients](https://github.com/indigo-iam/iam/blob/develop/iam-login-service/src/main/java/it/infn/mw/iam/config/TaskConfig.java#L152) task only when `openid-fed` is included among the active profiles.

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

rmiccoli committed:
Run expired client task when oidfed profile is active

rmiccoli commented:
> Remember also to run the [disableExpiredClients](https://github.com/indigo-iam/iam/blob/develop/iam-login-service/src/main/java/it/infn/mw/iam/config/TaskConfig.java#L152) task only when `openid-fed` is included among the active profiles.

Done :)

rmiccoli committed:
Fix sonar issues

rmiccoli committed:
Increase jwk size

rmiccoli committed:
Add a log during token request

rmiccoli committed:
Merge tag 'v1.13.1' into issue-1059

v1.13.1

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

rmiccoli committed:
Avoid string concatenation

rmiccoli committed:
Increase jwk size

to the maximum allowed by TEXT

rmiccoli committed:
Add request object signing alg in clientDTO

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

rmiccoli committed:
Add filter to log basic auth header

rmiccoli committed:
Move the doFilter outside the debug block

rmiccoli committed:
Extend test to check authz code flow works

and send client id and secret url-encoded
in basic auth header

rmiccoli committed:
Dump all headers

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

rmiccoli committed:
Fix code smell

rmiccoli committed:
Fix request object signing algorithm

also during client update

rmiccoli committed:
Replace string with boolean

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

rmiccoli committed:
Update client when expired

after trust chain re-validation.
Previously, if the client was expired, it was re-created
by violating the unique constraint on client_id, which
corresponds to the entity_id (URL) of the RP.

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/config/BasicAuthLoggingFilter.java:
```diff
@@ -0,0 +1,65 @@
+import org.springframework.stereotype.Component;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+@Component
```
Add `@Profile('openid-federation')`

rmiccoli replied:
Done

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

rmiccoli committed:
Enable basic auth filter only for oidfed

for debugging purposes

rmiccoli committed:
Disable Docker BuildKit

rmiccoli committed:
Improve workflow

rmiccoli committed:
Revert "Enable basic auth filter only for oidfed"

This reverts commit eeb5b98f26065183a04769883e195c995f660d21.

rmiccoli committed:
Resolve conflict

rmiccoli committed:
Restore previous maven file

rmiccoli committed:
Merge remote-tracking branch 'origin/develop' into issue-1059

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oidc/IamAuthorizationRequestFilter.java:
```diff
@@ -0,0 +1,479 @@
+  private final AutomaticClientRegistrationMapper clientMapper;
+
+  private LoginHintExtracter loginHintExtracter = new RemoveLoginHintsWithHTTP();
+  private RequestMatcher requestMatcher = new AntPathRequestMatcher("/authorize");
```
I'd suggest to convert it into a static final constant and to rename it into something like "authorizeMatcher"

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oidc/TrustChainService.java:
```diff
@@ -26,7 +25,6 @@
 import com.nimbusds.openid.connect.sdk.federation.trust.TrustChain;
 
 @Service
-@Profile("openid-federation")
```
This means we have always an instance of a TrustChain service. Even if `openid-federation` is not specified within the active profiles. Why? It is needed by others?

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oidc/TrustChainResolver.java:
Should it be added a `@Profile("openid-federation")` ? This resolver could not exist in the other cases

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/config/MitreServicesConfig.java:
```diff
@@ -187,21 +185,6 @@ ServerConfigInterceptor serverConfigInterceptor() {
-  }
-
-  @Bean(name = "mitreAuthzRequestFilter")
-  AuthorizationRequestFilter authorizationRequestFilter() {
```
In order to not add openid fed logic when no profile is added, I tend to keep the MitreID implementation when no openid-federation" profile is present.
Otherwise we can have this implementation.

Importing mitreID, refactor it and make something like fed one extends default one could be also done but it's a bigger change.

So something like:

```
  @Bean
  @Profile("!openid-federation")
  AuthorizationRequestFilter authorizationRequestFilter() {
     return new AuthorizationRequestFilter();
  }

  @Bean
  @Profile("openid-federation")
  AuthorizationRequestFilter authorizationRequestFilter(Environment env, IamClientRepository clientRepo, DefaultClientManagementService clientManagementService, RedirectResolver redirectResolver, TrustChainService trustChainService, AutomaticClientRegistrationMapper clientMapper) {
     return new IamAuthorizationRequestFilter(env, clientRepo, clientManagementService, redirectResolver, trustChainService, clientMapper);
  }
```

Evaluate to move the second bean inside IamConfig. But probably they should be kept one aside the other. So Mitre services can be a tolerated place XD
I don't think qualifiers are required.
This means also to remove the exclusion of AuthorizationRequestFilter in the main class.

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oidc/IamAuthorizationRequestFilter.java:
```diff
@@ -0,0 +1,479 @@
+    Optional<ClientDetailsEntity> client = Optional.empty();
+
+    String clientId = params.get(CLIENT_ID);
+    if (clientId != null) {
```
no else for clientId null, probably a sendAuthenticationError could be necessary.

enricovianello committed:
Extract the profile logic from AuthorizationRequestFilter

enricovianello committed:
Fix x509 session errors on test

If session is not present, create it

enricovianello committed:
Fix small Sonar issues

enricovianello committed:
Refactor federation exceptions
